qualtran/bloqs/chemistry/thc.py

```diff
@@ -0,0 +1,420 @@
+            mu=regs['mu'],
+            nu=regs['nu'],
+            succ=regs['succ'],
+            eq_nu_mp1=regs['eq_nu_mp1'],
+        )
+        # 2. Make contiguous register from mu and nu and store in register `s`.
+        mu, nu, s = bb.add(ToContiguousIndex(log_mu, log_d), mu=mu, nu=nu, s=s)
+        theta = regs['theta']
+        # 3. Load alt / keep values
+        qroam = CirqGateAsBloq(
+            SelectSwapQROM(
+                *(self.theta, self.alt_theta, self.alt_mu, self.alt_nu, self.keep),
+                target_bitsizes=(1, 1, alt_bitsize, alt_bitsize, self.keep_bitsize),
+            )
+        )
+        s, theta, alt_theta, alt_mu, alt_nu, keep = bb.add(
+            qroam,
+            selection=s,
+            target0=theta,
+            target1=alt_theta,
+            target2=alt_mu,
+            target3=alt_nu,
+            target4=keep,
+        )
+        sigma = bb.add(OnEach(self.keep_bitsize, Hadamard()), q=sigma)
+        lte_gate = CirqGateAsBloq(LessThanEqualGate(self.keep_bitsize, self.keep_bitsize))
+        keep, sigma, less_than = add_from_bloq_register_flat_qubits(
+            bb, lte_gate, keep=keep, sigma=sigma, less_than=less_than
+        )
+        cz = CirqGateAsBloq(cirq.ControlledGate(cirq.Z))
+        alt_theta, less_than = add_from_bloq_register_flat_qubits(
+            bb, cz, alt_theta=alt_theta, less_than=less_than
+        )
+        cz = CirqGateAsBloq(cirq.ControlledGate(cirq.Z, control_values=(0,)))
+        # negative control on the less_than register
+        less_than, theta = add_from_bloq_register_flat_qubits(
+            bb, cz, less_than=less_than, theta=theta
+        )
+        less_than, alt_mu, mu = bb.add(CSwapApprox(bitsize=log_mu), ctrl=less_than, x=alt_mu, y=mu)
+        less_than, alt_nu, nu = bb.add(CSwapApprox(bitsize=log_mu), ctrl=less_than, x=alt_nu, y=nu)
+        keep, sigma, less_than = add_from_bloq_register_flat_qubits(
+            bb, lte_gate, keep=keep, sigma=sigma, less_than=less_than
+        )
+        # Select expects two plus states so set them up here.
+        plus_a = bb.add(Hadamard(), q=regs['plus_a'])
+        plus_b = bb.add(Hadamard(), q=regs['plus_b'])
+        # Need a toffoli gate with one negative control. And it into an ancilla
+        # and control off this for the swaps.
+        junk = bb.allocate(1)
+        ctrls = bb.join(np.array([eq_nu_mp1, plus_a]))
+        ctrls, junk = bb.add(
+            CirqGateAsBloq(MultiControlPauli(cvs=(0, 1), target_gate=cirq.X)),
+            controls=ctrls,
+            target=junk,
+        )
+        eq_nu_mp1, plus_a = bb.split(ctrls)
+        junk, mu, nu = bb.add(CSwapApprox(bitsize=log_mu), ctrl=junk, x=mu, y=nu)
+        bb.free(bb.join(np.array([less_than, alt_theta])))
+        bb.free(junk)
+        bb.free(s)
+        bb.free(sigma)
+        bb.free(keep)
+        bb.free(alt_mu)
+        bb.free(alt_nu)
```

Comment:
what do you think about putting some of these `free` calls closer to the last time the variable is used?

Reply:
done.